Sources/RemoteMessaging/RemoteMessagingProcessing.swift

```diff
@@ -0,0 +1,101 @@
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import BrowserServicesKit
+import Common
+import Configuration
+import Foundation
+
+/**
+ * This protocol defines API for providing RMF config matcher
+ * that contains values of matched attributes that the config
+ * file is evaluated against.
+ *
+ * Client apps should implement it and pass to a class implementing
+ * RemoteMessagingProcessing.
+ */
+public protocol RemoteMessagingConfigMatcherProviding {
+    func refreshConfigMatcher(using store: RemoteMessagingStoring) async -> RemoteMessagingConfigMatcher
+}
+
+/**
+ * This protocol defines API for Remote Messaging client in the app.
+ */
+public protocol RemoteMessagingProcessing {
+    /// Defines endpoint URL where the config file is available.
+    var endpoint: URL { get }
+
+    /// This holds the fetcher that downloads the config file from the server.
+    var configurationFetcher: RemoteMessagingConfigFetching { get }
+
+    /// This holds the config matcher provider that updates the config matcher before the config is evaluated.
+    var configMatcherProvider: RemoteMessagingConfigMatcherProviding { get }
+
+    /// Provides feature flag support for RMF.
+    var remoteMessagingAvailabilityProvider: RemoteMessagingAvailabilityProviding { get }
+
+    /**
+     * This function returns a config processor.
+     *
+     * Config processor performs evaluation of the JSON config file against the matcher containing
+     * app-specific, device-specific and user-specific matched attributes. Default implementation is provided.
+     */
+    func configProcessor(for configMatcher: RemoteMessagingConfigMatcher) -> RemoteMessagingConfigProcessing
+
+    /**
+     * This is the entry point to RMF from the client app.
+     *
+     * This function fetches the config, updates config matcher, evaluates the config against the matcher
+     * and stores the result as needed. Client apps should call this function in order to refresh remote messages.
+     * When messages are updated, `RemoteMessagingStore.Notifications.remoteMessagesDidChange` notification is posted.
+     * Default implementation is provided.
+     */
+    func fetchAndProcess(using store: RemoteMessagingStoring) async throws
+}
+
+public extension RemoteMessagingProcessing {
+
+    func configProcessor(for configMatcher: RemoteMessagingConfigMatcher) -> RemoteMessagingConfigProcessing {
+        RemoteMessagingConfigProcessor(remoteMessagingConfigMatcher: configMatcher)
+    }
+
+    func fetchAndProcess(using store: RemoteMessagingStoring) async throws {
+        guard remoteMessagingAvailabilityProvider.isRemoteMessagingAvailable else {
+            os_log("Remote messaging feature flag is disabled, skipping fetching messages", log: .remoteMessaging, type: .debug)
+            return
+        }
+        do {
+            let statusResponse = try await configurationFetcher.fetchRemoteMessagingConfig()
```

Comment:
I think this might be some old naming, where this func used to return status code or something? But now it is config indeed, isn't it?

I would actually suggest a couple of helper func here to make this functionality here more readable, e.g:
private func fetchRemoteMessagingConfig() async throws -> RemoteMessagingConfig
private func processRemoteMessagingConfig(_ config: RemoteMessagingConfig, using store: RemoteMessagingStoring) async throws -> ProcessedResult

where first func just implements: try await configurationFetcher.fetchRemoteMessagingConfig()
second: pretty much old the rest

Reply:
Thanks! Yeah, I agree it's a bit messy. This function used to be called `fetchRemoteMessages`. Now I've renamed it but failed to notice that there are two `fetchRemoteMessagingConfig` functions being called here, from 2 different protocols 🤦  and they do 2 different things, first one fetches the config from the server, and the second one fetches config metadata from the local database. 😬 

Let me try to make it right :)